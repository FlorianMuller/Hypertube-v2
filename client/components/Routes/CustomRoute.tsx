import React, { ReactElement, ElementType } from "react";
import { Redirect, Route } from "react-router";

import useApi from "../../hooks/useApi";
import { Fixture, ApiAuthResponse } from "../../models/models";

import Loading from "./Loading";

interface Props {
  authComponent?: ElementType;
  notAuthComponent?: ElementType;
  exact?: boolean;
  path?: string;
  fixture?: Fixture;
}

const CustomRoute = ({
  authComponent: AuthComponent,
  notAuthComponent: NotAuthComponent,
  path,
  exact,
  fixture
}: Props): ReactElement => {
  const { data, loading, error } = useApi<ApiAuthResponse, ApiAuthResponse>(
    "/check-auth",
    fixture
  );

  return (
    <Route
      path={path}
      exact={exact}
      render={(): ReactElement => {
        if (error?.response?.status >= 500 && error.response.status <= 599)
          return <div>Error</div>;
        if (loading) return <Loading />;

        if (AuthComponent && data?.validToken) return <AuthComponent />;
        if (NotAuthComponent && !data?.validToken) return <NotAuthComponent />;
        if (!loading) return <Redirect to="/" />;

        return null;
      }}
    />
  );
};

export default CustomRoute;
