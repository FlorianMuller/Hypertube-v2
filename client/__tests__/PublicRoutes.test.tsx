import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import { Router } from 'react-router-dom';

import CustomRoute from '../components/Routes/CustomRoute';
import history from '../helpers/history';

interface Props {
  children: ReactElement;
}

const MockComponent = ({ children }: Props): ReactElement => (
  <Router history={history}>
    {children}
  </Router>
);

describe('CustomRoute', () => {
  it('Is connected, private route : show div', () => {
    const tree = renderer
      .create(
        <MockComponent>
          <CustomRoute
            exact
            component={(): ReactElement => <div>Is connected, private route : show div</div>}
            requireAuth
            path="/"
            fixture={{
              loading: false,
              error: null,
              data: { validToken: true },
            }}
          />
        </MockComponent>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Is connected, public route : hide div', () => {
    const tree = renderer
      .create(
        <MockComponent>
          <CustomRoute
            exact
            component={(): ReactElement => <div>Is connected, public route : hide div</div>}
            requireAuth={false}
            path="/"
            fixture={{
              loading: false,
              error: null,
              data: { validToken: true },
            }}
          />
        </MockComponent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Is not connected, private route : hide div', () => {
    const tree = renderer
      .create(
        <MockComponent>
          <CustomRoute
            exact
            component={(): ReactElement => <div>Is not connected, private route : hide div</div>}
            requireAuth
            path="/"
            fixture={{
              loading: false,
              error: null,
              data: { validToken: false },
            }}
          />
        </MockComponent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Is not connected, public route : show div', () => {
    const tree = renderer
      .create(
        <MockComponent>
          <CustomRoute
            exact
            component={(): ReactElement => <div>Is not connected, public route : show div</div>}
            requireAuth={false}
            path="/"
            fixture={{
              loading: false,
              error: null,
              data: { validToken: true },
            }}
          />
        </MockComponent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Loading', () => {
    const tree = renderer
      .create(
        <MockComponent>
          <CustomRoute
            exact
            component={(): ReactElement => <div>Loading</div>}
            requireAuth={false}
            path="/"
            fixture={{
              loading: true,
              error: null,
              data: { validToken: true },
            }}
          />
        </MockComponent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Error', () => {
    const tree = renderer
      .create(
        <MockComponent>
          <CustomRoute
            exact
            component={(): ReactElement => <div>Error</div>}
            requireAuth={false}
            path="/"
            fixture={{
              loading: false,
              error: {
                error: true,
              },
              data: { validToken: true },
            }}
          />
        </MockComponent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
