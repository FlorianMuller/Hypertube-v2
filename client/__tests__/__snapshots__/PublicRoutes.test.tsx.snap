// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomRoute Error 1`] = `
<div>
  Error
</div>
`;

exports[`CustomRoute Is connected, private route : show div 1`] = `
<div>
  Is connected, private route : show div
</div>
`;

exports[`CustomRoute Is connected, public route : hide div 1`] = `
<div>
  Is connected, public route : hide div
</div>
`;

exports[`CustomRoute Is not connected, private route : hide div 1`] = `null`;

exports[`CustomRoute Is not connected, public route : show div 1`] = `
<div>
  Is not connected, public route : show div
</div>
`;

exports[`CustomRoute Loading 1`] = `
<div>
  Loading
</div>
`;
